### PR TITLE
use new form system, compatible with igor+

### LINKED ISSUE #14: Form Events and Compatibility with upcoming release Igor
This plugin seems to make use of one or more form related action events that will be removed in the next DokuWiki release. Please have a look at https://www.patreon.com/posts/better-forms-58551930 for details.

My apologies if this has already been addressed in your code. If you have any questions on how to implement the needed changes please let me know.

### PR DESCRIPTION
**Pull request by schplurtz**
HI,

This PR makes use of the new form and event system of the upcoming DokuWiki Igor version. The plugin will still be compatible with previous DW versions.

If you want to give it a try before you merge, visit https://aih8uw8shahshieti.bidule.https443.net/. This temporary DW will be automatically destroyed in a few weeks.

This should fix #14.

### CODE REVIEW

File: action.php [hunk 1]
```diff
@@ -144,9 +145,39 @@ public function handleHtmlEditformOutput($event, $param) {
...
-        $form = & $event->data;
-        $pos = $form->findElementByType('wikitext');
-        if ($pos===false) return;
+        $form = $event->data;
+        if (is_a($form, \dokuwiki\Form\Form::class)) {
+            $pos = $form->findPositionByType('textarea'); // new
+            if ($pos===false) return;
+            $this->handleHtmlEditformOutputNG($form, $pos, $origtext);
+        } else {
+            $pos = $form->findElementByType('wikitext'); // old
```
Comment (zuiderkwast) [inline action.php:154]: Here too. (See other comment regarding "old" and "new".)

File: action.php [hunk 2]
```diff
@@ -144,9 +145,39 @@ public function handleHtmlEditformOutput($event, $param) {
...
+        // Before the wikitext...
+        $form->addElement(new dokuwiki\Form\TagOpenElement('div', array('id'=>'wrapper__wikitext','class'=>'hor')), $pos++);
+        // After the wikitext...
+        $pos++;
+        $form->addElement(new dokuwiki\Form\TagCloseElement('div'), $pos++);
+        $form->addElement(new dokuwiki\Form\TagOpenElement('div', array('id'=>'wrapper__sourcetext','class'=>'hor')), $pos++);
+        $origelem = '<textarea id="translate__sourcetext" '.
+                    //buildAttributes($attrs,true).
+                    'class="edit" readonly="readonly" cols="80" rows="10"'.
+                    'style="width:100%;"'.//  height:600px; overflow:auto
```
Comment (zuiderkwast) [inline action.php:170]: What about height and overflow? Something you tried and you're not sure they're needed?

I got this when I click "edit" without logging in:

image

File: action.php [hunk 3]
```diff
@@ -144,9 +145,39 @@ public function handleHtmlEditformOutput($event, $param) {
...
+                    '>'.NL.
+                    hsc($origtext).
+                    '</textarea>';
+        $form->addElement(new dokuwiki\Form\HTMLElement($origelem), $pos++);
+        $form->addElement(new dokuwiki\Form\TagCloseElement('div'), $pos++);
+        $form->addElement(new dokuwiki\Form\TagOpenElement('div', array('class'=>'clearer')), $pos++);
+        $form->addElement(new dokuwiki\Form\TagCloseElement('div'), $pos++);
+    }
+
+    protected function handleHtmlEditformOutputLegacy($form, $pos, $origtext) {
```
Comment (zuiderkwast) [inline action.php:180]: Perhaps add a comment that this is used for dokuwiki < x.y.z (version)

**Review by zuiderkwast — COMMENTED** [review]
Nice!

I tested it and it seems to work very well! Thanks for fixing this and for supplying this test site!

File: action.php (hunk 3 above)
Comment (schplurtz) [reply action.php:180]: noted. will be fixed this evening

File: action.php (hunk 1 above)
Comment (schplurtz) [reply action.php]: done too.

File: action.php (hunk 3 above)
Comment (schplurtz) [reply action.php:192]: done again

**Comment by schplurtz** [comment]
> 



> I tested it and it seems to work very well! Thanks for fixing this and for supplying this test site!

That's working. But it seems I did not do it the DokuWiki way. When I was looking at how DokuWiki creates its textarea I discovered that the Form class has methods to add elements and it's not necessary to manually create all those `dokuwiki\Form\Foo` elements. See https://github.com/splitbrain/dokuwiki/blob/f843335e38f43e39981e329010ab831750eff4a3/inc/Ui/Editor.php#L208-L209

I'll add another commit to create the elements "the DokuWiki way".

File: action.php (hunk 2 above)
Comment (zuiderkwast) [reply action.php:179]: OK, that's fine.

Too much CSS can mess up user stylesheets too, so better leave it as it is.

I just saw "//  height:600px; overflow:auto" commented out, and "//buildAttributes($attrs,true)" too. That's why I mentioned overflow. Perhaps you can delete those if they're not used.

Comment (schplurtz) [reply action.php]: > Perhaps you can delete those if they're not used.

he he, that's easily done. Anyway, the last commit (d238a5d61e4271fb721ca00eeaec74a41d9f6c23) made this disappear

Comment (schplurtz) [reply action.php]: I think I've resolved the problem. This is no CSS problem, but JS. your JS which is supposed to take care of this is outdated. On current DW, it crashes.

CSS is incomprehensible black magic to me, JS is only a strange but still manageable language. I'll propose a solution for this in another PR.

image

This was reported as #15.

**Review by zuiderkwast — APPROVED** [review]
Nice!

Is this ready to merge?

**Comment by schplurtz** [comment]
> Is this ready to merge?

Sorry, all those commits made it unclear. but yes, this is ready now.

**Pull request merged**